File: iosApp/iosApp/Features/ComfyUI/ComfyUIGenerationViewModel.swift
import Foundation
import Shared
import UIKit

/// KMP seed value meaning "random"; shown as an empty seed field.
private let randomSeed: Int64 = -1

@MainActor
final class ComfyUIGenerationViewModelOwner: ObservableObject {
    let vm: Feature_comfyuiComfyUIGenerationViewModel
    private let store = ViewModelStore()

    // Published properties mapped from KMP UiState
    @Published var checkpoints: [String] = []
    @Published var selectedCheckpoint = ""
    @Published var prompt = ""
    @Published var negativePrompt = ""
    @Published var steps: Double = 20
    @Published var cfgScale: Double = 7.0
    @Published var width: String = "512"
    @Published var height: String = "512"
    @Published var seed: String = ""
    @Published var isLoadingCheckpoints = false
    // LoRA
    @Published var availableLoras: [String] = []
    @Published var loraSelections: [LoraSelection] = []
    // ControlNet
    @Published var availableControlNets: [String] = []
    @Published var controlNetEnabled = false
    @Published var selectedControlNet = ""
    @Published var controlNetStrength: Double = 1.0
    // Custom workflow
    @Published var customWorkflowJson: String?
    @Published var workflowImportError: String?
    // Dynamic workflow parameters
    @Published var extractedParameters: [Feature_comfyuiExtractedParameter] = []
    @Published var isLoadingParameters = false
    // Inpainting mask
    @Published var initImageFilename: String?
    @Published var maskImageFilename: String?
    @Published var denoiseStrength: Double = 0.75
    // Generation
    @Published var generationStatus: GenerationStatus = .idle
    @Published var currentStep: Int32 = 0
    @Published var totalSteps: Int32 = 0
    @Published var resultImageUrls: [String] = []
    @Published var error: String?
    @Published var imageSaveSuccess: KotlinBoolean?
    @Published var previewImage: UIImage?
    @Published var currentNodeName: String = ""

    // KMP values last mirrored into the numeric text fields. A field is rewritten only when KMP
    // changes its value, so an unrelated emission never snaps a cleared or half-typed field back.
    private var mirroredWidth: Int32?
    private var mirroredHeight: Int32?
    private var mirroredSeed: Int64?

    var progressFraction: Float {
        guard totalSteps > 0 else { return 0 }
        return Float(currentStep) / Float(totalSteps)
    }

    /// Width/height text that does not parse is never forwarded, so KMP would still submit the last valid size.
    var hasValidDimensions: Bool {
        Int32(width) != nil && Int32(height) != nil
    }

    init() {
        vm = KoinHelper.shared.createComfyUIGenerationViewModel()
        store.put(key: "ComfyUIGenerationViewModel", viewModel: vm)
    }

    deinit { store.clear() }

    func observeUiState() async {
        for await state in vm.uiState {
            checkpoints = state.checkpoints as? [String] ?? []
            selectedCheckpoint = state.selectedCheckpoint
            prompt = state.prompt
            negativePrompt = state.negativePrompt
            steps = Double(state.steps)
            cfgScale = state.cfgScale
            mirrorNumericFields(width: state.width, height: state.height, seed: state.seed)
            isLoadingCheckpoints = state.isLoadingCheckpoints
            availableLoras = state.availableLoras as? [String] ?? []
            loraSelections = state.loraSelections as? [LoraSelection] ?? []
            availableControlNets = state.availableControlNets as? [String] ?? []
            controlNetEnabled = state.controlNetEnabled
            selectedControlNet = state.selectedControlNet
            controlNetStrength = Double(state.controlNetStrength)
            customWorkflowJson = state.customWorkflowJson
            workflowImportError = state.workflowImportError
            extractedParameters = state.extractedParameters as? [Feature_comfyuiExtractedParameter] ?? []
            isLoadingParameters = state.isLoadingParameters
            initImageFilename = state.initImageFilename
            maskImageFilename = state.maskImageFilename
            denoiseStrength = state.denoiseStrength
            generationStatus = state.generationStatus
            currentStep = state.currentStep
            totalSteps = state.totalSteps
            error = state.error
            imageSaveSuccess = state.imageSaveSuccess
            currentNodeName = state.currentNodeName
            // Preview image from bytes
            if let bytes = state.previewImageBytes {
                let data = kotlinByteArrayToData(bytes)
                if let image = UIImage(data: data) {
                    previewImage = image
                }
            }
            // Result image URLs
            if let result = state.result {
                resultImageUrls = result.imageUrls
            }
        }
    }

    // MARK: - Actions (delegate to KMP VM)

    func onCheckpointSelected(_ checkpoint: String) {
        selectedCheckpoint = checkpoint
        vm.onCheckpointSelected(checkpoint: checkpoint)
    }
    func onPromptChanged(_ value: String) {
        prompt = value
        vm.onPromptChanged(prompt: value)
    }
    func onNegativePromptChanged(_ value: String) {
        negativePrompt = value
        vm.onNegativePromptChanged(prompt: value)
    }
    /// Rounds so the slider, its label and the submitted step count stay the same integer.
    func onStepsChanged(_ steps: Double) {
        let rounded = steps.rounded()
        self.steps = rounded
        vm.onStepsChanged(steps: Int32(rounded))
    }
    func onCfgScaleChanged(_ cfg: Double) {
        cfgScale = cfg
        vm.onCfgScaleChanged(cfg: cfg)
    }
    func onWidthChanged(_ width: String) {
        self.width = width
        if let w = Int32(width) { vm.onWidthChanged(width: w) }
    }
    func onHeightChanged(_ height: String) {
        self.height = height
        if let h = Int32(height) { vm.onHeightChanged(height: h) }
    }
    /// Empty or unparseable text means a random seed.
    func onSeedChanged(_ seed: String) {
        self.seed = seed
        vm.onSeedChanged(seed: Int64(seed) ?? randomSeed)
    }
    func onLoraAdded(_ name: String) { vm.onLoraAdded(loraName: name) }
    func onLoraRemoved(_ name: String) { vm.onLoraRemoved(loraName: name) }
    func onLoraStrengthChanged(name: String, strengthModel: Float, strengthClip: Float) {
        vm.onLoraStrengthChanged(loraName: name, strengthModel: strengthModel, strengthClip: strengthClip)
    }
    func onControlNetToggled(_ enabled: Bool) {
        controlNetEnabled = enabled
        vm.onControlNetToggled(enabled: enabled)
    }
    func onControlNetSelected(_ model: String) {
        selectedControlNet = model
        vm.onControlNetSelected(model: model)
    }
    func onControlNetStrengthChanged(_ strength: Double) {
        controlNetStrength = strength
        vm.onControlNetStrengthChanged(strength: Float(strength))
    }
    func onImportWorkflow(_ json: String) { vm.onImportWorkflow(jsonInput: json) }
    func onClearCustomWorkflow() { vm.onClearCustomWorkflow() }
    func onParameterValueChanged(nodeId: String, paramName: String, newValue: String) {
        vm.onParameterValueChanged(nodeId: nodeId, paramName: paramName, newValue: newValue)
    }
    func onRefreshParameters() { vm.onRefreshParameters() }
    func onMaskUploaded(_ filename: String) { vm.onMaskUploaded(filename: filename) }
    func onClearMask() { vm.onClearMask() }
    func onDenoiseStrengthChanged(_ strength: Double) {
        denoiseStrength = strength
        vm.onDenoiseStrengthChanged(strength: strength)
    }
    func onGenerate() { vm.onGenerate() }
    func onSaveImage(url: String) { vm.onSaveImage(imageUrl: url) }
    func onDismissSaveResult() { vm.onDismissSaveResult() }
    func onInterrupt() { vm.onInterrupt() }

    private func mirrorNumericFields(width newWidth: Int32, height newHeight: Int32, seed newSeed: Int64) {
        if newWidth != mirroredWidth, Int32(width) != newWidth { width = String(newWidth) }
        if newHeight != mirroredHeight, Int32(height) != newHeight { height = String(newHeight) }
        if newSeed != mirroredSeed, seedValue(of: seed) != newSeed {
            seed = newSeed == randomSeed ? "" : String(newSeed)
        }
        mirroredWidth = newWidth
        mirroredHeight = newHeight
        mirroredSeed = newSeed
    }
}

private func seedValue(of text: String) -> Int64? {
    text.isEmpty ? randomSeed : Int64(text)
}

/// Convert Kotlin ByteArray to Swift Data
private func kotlinByteArrayToData(_ byteArray: KotlinByteArray) -> Data {
    let size = byteArray.size
    var bytes = [UInt8](repeating: 0, count: Int(size))
    for i in 0..<size {
        bytes[Int(i)] = UInt8(bitPattern: byteArray.get(index: i))
    }
    return Data(bytes)
}

File: feature/feature-comfyui/src/commonMain/kotlin/com/riox432/civitdeck/feature/comfyui/presentation/ComfyUIGenerationViewModel.kt
package com.riox432.civitdeck.feature.comfyui.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.riox432.civitdeck.data.image.SaveGeneratedImageUseCase
import com.riox432.civitdeck.domain.model.ComfyUIGenerationParams
import com.riox432.civitdeck.domain.model.GenerationResult
import com.riox432.civitdeck.domain.model.GenerationStatus
import com.riox432.civitdeck.domain.model.LoraSelection
import com.riox432.civitdeck.domain.repository.ComfyUIConnectionRepository
import com.riox432.civitdeck.feature.comfyui.domain.model.ExtractedParameter
import com.riox432.civitdeck.feature.comfyui.domain.usecase.ExtractWorkflowParametersUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.FetchComfyUICheckpointsUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.FetchComfyUIControlNetsUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.FetchComfyUILorasUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.FetchObjectInfoUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.ImportWorkflowUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.InjectWorkflowParametersUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.InterruptComfyUIGenerationUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.ObserveGenerationProgressUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.PollComfyUIResultUseCase
import com.riox432.civitdeck.feature.comfyui.domain.usecase.SubmitComfyUIGenerationUseCase
import com.riox432.civitdeck.util.Logger
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class GenerationUiState(
    val checkpoints: List<String> = emptyList(),
    val selectedCheckpoint: String = "",
    val prompt: String = "",
    val negativePrompt: String = "",
    val steps: Int = 20,
    val cfgScale: Double = 7.0,
    val width: Int = 512,
    val height: Int = 512,
    val seed: Long = -1,
    val samplerName: String = "euler",
    val scheduler: String = "normal",
    val isLoadingCheckpoints: Boolean = false,
    // LoRA
    val availableLoras: List<String> = emptyList(),
    val isLoadingLoras: Boolean = false,
    val loraSelections: List<LoraSelection> = emptyList(),
    // ControlNet
    val availableControlNets: List<String> = emptyList(),
    val isLoadingControlNets: Boolean = false,
    val controlNetEnabled: Boolean = false,
    val selectedControlNet: String = "",
    val controlNetStrength: Float = 1.0f,
    // Custom workflow
    val customWorkflowJson: String? = null,
    val workflowImportError: String? = null,
    // Dynamic workflow parameters
    val extractedParameters: List<ExtractedParameter> = emptyList(),
    val isLoadingParameters: Boolean = false,
    // Inpainting mask
    val initImageFilename: String? = null,
    val maskImageFilename: String? = null,
    val denoiseStrength: Double = 0.75,
    // Generation
    val generationStatus: GenerationStatus = GenerationStatus.Idle,
    val currentStep: Int = 0,
    val totalSteps: Int = 0,
    val result: GenerationResult? = null,
    val error: String? = null,
    val imageSaveSuccess: Boolean? = null,
    /** Latest preview image bytes from WebSocket during generation. */
    val previewImageBytes: ByteArray? = null,
    val currentNodeName: String = "",
) {
    /** Progress fraction in [0f, 1f]. Returns 0 when totalSteps is unknown. */
    val progressFraction: Float
        get() = if (totalSteps > 0) currentStep.toFloat() / totalSteps.toFloat() else 0f
}

@Suppress("LongParameterList", "TooManyFunctions")
class ComfyUIGenerationViewModel(
    private val fetchCheckpoints: FetchComfyUICheckpointsUseCase,
    private val fetchLoras: FetchComfyUILorasUseCase,
    private val fetchControlNets: FetchComfyUIControlNetsUseCase,
    private val importWorkflow: ImportWorkflowUseCase,
    private val extractParameters: ExtractWorkflowParametersUseCase,
    private val injectParameters: InjectWorkflowParametersUseCase,
    private val fetchObjectInfo: FetchObjectInfoUseCase,
    submitGeneration: SubmitComfyUIGenerationUseCase,
    pollResult: PollComfyUIResultUseCase,
    observeProgress: ObserveGenerationProgressUseCase,
    interruptGeneration: InterruptComfyUIGenerationUseCase,
    saveImage: SaveGeneratedImageUseCase,
    repository: ComfyUIConnectionRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow(GenerationUiState())
    val uiState: StateFlow<GenerationUiState> = _uiState

    private val generationDelegate = GenerationExecutionDelegate(
        scope = viewModelScope,
        uiState = _uiState,
        submitGeneration = submitGeneration,
        pollResult = pollResult,
        observeProgress = observeProgress,
        interruptGeneration = interruptGeneration,
        saveImage = saveImage,
        repository = repository,
    )

    init {
        loadCheckpoints()
        loadLoras()
        loadControlNets()
    }

    private fun loadCheckpoints() {
        _uiState.update { it.copy(isLoadingCheckpoints = true) }
        launchWithErrorHandling(
            tag = "Failed to load checkpoints",
            onError = { e -> _uiState.update { it.copy(isLoadingCheckpoints = false, error = e.message) } },
        ) {
            val list = fetchCheckpoints()
            _uiState.update {
                it.copy(
                    checkpoints = list,
                    selectedCheckpoint = list.firstOrNull() ?: "",
                    isLoadingCheckpoints = false,
                )
            }
        }
    }

    private fun loadLoras() {
        _uiState.update { it.copy(isLoadingLoras = true) }
        launchWithErrorHandling(
            tag = "Failed to fetch loras",
            onError = { _uiState.update { it.copy(isLoadingLoras = false) } },
        ) {
            val list = fetchLoras()
            _uiState.update { it.copy(availableLoras = list, isLoadingLoras = false) }
        }
    }

    private fun loadControlNets() {
        _uiState.update { it.copy(isLoadingControlNets = true) }
        launchWithErrorHandling(
            tag = "Failed to fetch control nets",
            onError = { _uiState.update { it.copy(isLoadingControlNets = false) } },
        ) {
            val list = fetchControlNets()
            _uiState.update { it.copy(availableControlNets = list, isLoadingControlNets = false) }
        }
    }

    fun onCheckpointSelected(checkpoint: String) {
        _uiState.update { it.copy(selectedCheckpoint = checkpoint) }
    }

    fun onPromptChanged(prompt: String) {
        _uiState.update { it.copy(prompt = prompt) }
    }

    fun onNegativePromptChanged(prompt: String) {
        _uiState.update { it.copy(negativePrompt = prompt) }
    }

    fun onStepsChanged(steps: Int) {
        _uiState.update { it.copy(steps = steps) }
    }

    fun onCfgScaleChanged(cfg: Double) {
        _uiState.update { it.copy(cfgScale = cfg) }
    }

    fun onWidthChanged(width: Int) {
        _uiState.update { it.copy(width = width) }
    }

    fun onHeightChanged(height: Int) {
        _uiState.update { it.copy(height = height) }
    }

    fun onSeedChanged(seed: Long) {
        _uiState.update { it.copy(seed = seed) }
    }

    // -- LoRA --

    fun onLoraAdded(loraName: String) {
        val current = _uiState.value.loraSelections
        if (current.none { it.name == loraName }) {
            _uiState.update { it.copy(loraSelections = current + LoraSelection(loraName)) }
        }
    }

    fun onLoraRemoved(loraName: String) {
        _uiState.update { state ->
            state.copy(loraSelections = state.loraSelections.filter { it.name != loraName })
        }
    }

    fun onLoraStrengthChanged(loraName: String, strengthModel: Float, strengthClip: Float) {
        _uiState.update { state ->
            state.copy(
                loraSelections = state.loraSelections.map {
                    if (it.name == loraName) {
                        it.copy(strengthModel = strengthModel, strengthClip = strengthClip)
                    } else {
                        it
                    }
                }
            )
        }
    }

    // -- ControlNet --

    fun onControlNetToggled(enabled: Boolean) {
        _uiState.update { it.copy(controlNetEnabled = enabled) }
    }

    fun onControlNetSelected(model: String) {
        _uiState.update { it.copy(selectedControlNet = model) }
    }

    fun onControlNetStrengthChanged(strength: Float) {
        _uiState.update { it.copy(controlNetStrength = strength) }
    }

    // -- Custom workflow --

    fun onImportWorkflow(jsonInput: String) {
        try {
            val validated = importWorkflow(jsonInput)
            _uiState.update { it.copy(customWorkflowJson = validated, workflowImportError = null) }
            extractWorkflowParameters(validated)
        } catch (e: IllegalStateException) {
            _uiState.update { it.copy(workflowImportError = e.message) }
        }
    }

    fun onClearCustomWorkflow() {
        _uiState.update {
            it.copy(
                customWorkflowJson = null,
                workflowImportError = null,
                extractedParameters = emptyList(),
            )
        }
    }

    // -- Dynamic workflow parameters --

    fun onParameterValueChanged(nodeId: String, paramName: String, newValue: String) {
        _uiState.update { state ->
            state.copy(
                extractedParameters = state.extractedParameters.map { param ->
                    if (param.nodeId == nodeId && param.paramName == paramName) {
                        param.copy(currentValue = newValue)
                    } else {
                        param
                    }
                }
            )
        }
    }

    fun onRefreshParameters() {
        val json = _uiState.value.customWorkflowJson ?: return
        extractWorkflowParameters(json)
    }

    private fun extractWorkflowParameters(workflowJson: String) {
        _uiState.update { it.copy(isLoadingParameters = true) }
        launchWithErrorHandling(
            tag = "Failed to extract workflow parameters",
            onError = { _uiState.update { it.copy(isLoadingParameters = false) } },
        ) {
            val objectInfoJson = fetchObjectInfo()
            val params = extractParameters(workflowJson, objectInfoJson)
            _uiState.update { it.copy(extractedParameters = params, isLoadingParameters = false) }
        }
    }

    // -- Inpainting mask --

    fun onInitImageUploaded(filename: String) {
        _uiState.update { it.copy(initImageFilename = filename) }
    }

    fun onMaskUploaded(filename: String) {
        _uiState.update { it.copy(maskImageFilename = filename) }
    }

    fun onClearMask() {
        _uiState.update {
            it.copy(initImageFilename = null, maskImageFilename = null)
        }
    }

    fun onDenoiseStrengthChanged(strength: Double) {
        _uiState.update { it.copy(denoiseStrength = strength) }
    }

    // -- Generation (delegated) --

    fun onGenerate() {
        val state = _uiState.value
        val hasCustomWorkflow = state.customWorkflowJson != null
        if (!hasCustomWorkflow && (state.selectedCheckpoint.isBlank() || state.prompt.isBlank())) return
        generationDelegate.onGenerate(buildParams(state))
    }

    fun onSaveImage(imageUrl: String) = generationDelegate.onSaveImage(imageUrl)

    fun onDismissSaveResult() = generationDelegate.onDismissSaveResult()

    fun onInterrupt() = generationDelegate.onInterrupt()

    private fun buildParams(state: GenerationUiState): ComfyUIGenerationParams {
        // If custom workflow has extracted parameters, inject modified values before submission
        val finalWorkflowJson = if (state.customWorkflowJson != null && state.extractedParameters.isNotEmpty()) {
            injectParameters(state.customWorkflowJson, state.extractedParameters)
        } else {
            state.customWorkflowJson
        }

        return ComfyUIGenerationParams(
            checkpoint = state.selectedCheckpoint,
            prompt = state.prompt,
            negativePrompt = state.negativePrompt,
            steps = state.steps,
            cfgScale = state.cfgScale,
            seed = state.seed,
            width = state.width,
            height = state.height,
            samplerName = state.samplerName,
            scheduler = state.scheduler,
            loraSelections = state.loraSelections,
            controlNetEnabled = state.controlNetEnabled,
            controlNetModel = state.selectedControlNet,
            controlNetStrength = state.controlNetStrength,
            customWorkflowJson = finalWorkflowJson,
            initImageFilename = state.initImageFilename,
            maskImageFilename = state.maskImageFilename,
            denoiseStrength = state.denoiseStrength,
        )
    }

    private inline fun launchWithErrorHandling(
        tag: String,
        crossinline onError: (Exception) -> Unit,
        crossinline block: suspend () -> Unit,
    ) {
        viewModelScope.launch {
            try {
                block()
            } catch (e: CancellationException) {
                throw e
            } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
                Logger.e(TAG, "$tag: ${e.message}")
                onError(e)
            }
        }
    }

    companion object {
        private const val TAG = "ComfyUIGenerationVM"
    }
}
